In [1]:
#Import Packages

#To build a NN model
from keras.models import Sequential

#To build a densely connected NN layer
from keras.layers import Dense

#Regularization Function
from keras.layers import Dropout

#Activation Function
from keras.layers import Activation

#Optimizer - Stochastic Gradient Descent
from keras.optimizers import SGD

#To convert integer targets to categorical targets#To build a NN model
from keras.models import Sequential

#To build a densely connected NN layer
from keras.layers import Dense

#Regularization Function
from keras.layers import Dropout
#Categorical Targets - N dimension vector which contains all zeros except for a 1 at a corresponding sample index
from keras.utils.np_utils import to_categorical

#Dataset - Handwritten images
from keras.datasets import mnist


Using Theano backend.


In [2]:
(x_train, y_train), (x_test, y_test) = mnist.load_data()

In [3]:
#x_train contains 60000 28*28 grayscale images of 10 digits(0-9).
print x_train.shape

#y_train contains their corresponding digit labels.
print y_train.shape

#x_test contains 10000 28*28 grayscale images of 10 digits(0-9).
print x_test.shape
#y_train contains their corresponding digit labels.
print y_test.shape

(60000, 28, 28)
(60000,)
(10000, 28, 28)
(10000,)


In [4]:
#Convert X_train, y_train, X_test and y_test in a format that can be fed into a DNN.
X_train = x_train.reshape(60000, 784)
Y_train = to_categorical(y_train, 10)
X_test = x_test.reshape(10000, 784)
Y_test = to_categorical(y_test, 10)

In [5]:
print X_train.shape
print y_train.shape
print X_test.shape
print y_test.shape

(60000, 784)
(60000,)
(10000, 784)
(10000,)


In [6]:
#Lets print out first sample which happens to be digit 7.

#Pixel values represent the brightness of the image, its value ranges from 0-255. 
#Typically 0 is taken to be white and 255 is taken to be black.
print "Pixel values of input digit image:"
print X_train[0]
print "Target digit value :" 
print y_test[0]

Pixel values of input digit image:
[  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
   0   0   0   0   0   0   0   0   3  18  18  18 126 136 175  26 166 255
 247 127   0   0   0   0   0   0   0   0   0   0   0   0  30  36  94 154
 170 253 253 253 253 253 225 172 253 242 195  64   0   0   0   0   0   0
   0   0   0   0   0  49 238 253 253 253 253 253 253 253 253 251  93  82
  82  56  39   0   0   0   0   0   0   0   0   0   0   0   0  18 219 253
 253 253 253 253

In [7]:
#Build NN model
model = Sequential()

#Using Dense add input layer and hidden layer.
#In this example - 
    #Input layer contains 784 input units ie) one input unit for each pixel value.
    #Hidden layer contains 25 (arbitrarily chosen) hidden units.
    #Apply sigmoid activation function in hidden layer. 
model.add(Dense(25, input_dim=784, activation='sigmoid'))

#Dropout helps to prevent overfitting
model.add(Dropout(0.2))

#Using Dense add output layer which contains 10 output units - one per each digit (0-9).
#Apply softmax activation function in output layer.
model.add(Dense(10,  activation='softmax'))
model.summary()


_________________________________________________________________
Layer (type)                 Output Shape              Param #   
dense_1 (Dense)              (None, 25)                19625     
_________________________________________________________________
dropout_1 (Dropout)          (None, 25)                0         
_________________________________________________________________
dense_2 (Dense)              (None, 10)                260       
Total params: 19,885
Trainable params: 19,885
Non-trainable params: 0
_________________________________________________________________


**Model Parameters**: In our example total params is **19885**. Lets see how its calculated.

Total Params = 19885 = 19625 + 260

 **19625** = (784 * 25) + 25
 
 where 
 784 = Input Units, 25 = Hidden Units and 25 = Bias Units
 
 **260** = (25 * 10) + 10
 
 where
 25 = Hidden Units, 10 = Output Units and 10 = Bias Units

In [8]:
#Configure the learning process.
model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=["accuracy"])

In [9]:
#Train the NN model for 10 epochs and after training the model with a batch_size of 32 training samples, perform parameter update.
model.fit(X_train, Y_train, epochs=10, batch_size=32, verbose=1)

Epoch 1/10
60000/60000 [==============================] - 2s - loss: 0.9449 - acc: 0.7267     
Epoch 2/10
60000/60000 [==============================] - 2s - loss: 0.5453 - acc: 0.8389     
Epoch 3/10
60000/60000 [==============================] - 2s - loss: 0.4966 - acc: 0.8544     
Epoch 4/10
60000/60000 [==============================] - 2s - loss: 0.4683 - acc: 0.8634     
Epoch 5/10
60000/60000 [==============================] - 2s - loss: 0.4514 - acc: 0.8679     - ETA: 2s - loss: 0.4229 - acc: 0
Epoch 6/10
60000/60000 [==============================] - 2s - loss: 0.4498 - acc: 0.8704     
Epoch 7/10
60000/60000 [==============================] - 2s - loss: 0.4403 - acc: 0.8735     
Epoch 8/10
60000/60000 [==============================] - 2s - loss: 0.4291 - acc: 0.8775     
Epoch 9/10
60000/60000 [==============================] - 2s - loss: 0.4278 - acc: 0.8775     
Epoch 10/10
60000/60000 [==============================] - 2s - loss: 0.4225 - acc: 0.8790     


In [10]:
#Compute the loss on test data.
score,acc = model.evaluate(X_test, Y_test, verbose=0)
#Print the accuracy
print '\nTest accuracy:%f %%' %(acc * 100)


Test accuracy:90.570000 %


In [11]:
#Lets now perform prediction
predicted_classes = model.predict_classes(X_test, verbose=0)
#Print target values and its predicted values.
print 'Target Num:%d, Output Num:%d' %(y_test[0], predicted_classes[0])
print 'Target Num:%d, Output Num:%d' %(y_test[99], predicted_classes[99])
print 'Target Num:%d, Output Num:%d' %(y_test[9999], predicted_classes[9999])

Target Num:7, Output Num:7
Target Num:9, Output Num:9
Target Num:6, Output Num:6
